plugins/hls-explicit-imports-plugin/src/Ide/Plugin/ExplicitImports.hs

```diff
@@ -187,49 +240,70 @@ instance NFData MinimalImports
 
 type instance RuleResult MinimalImports = MinimalImportsResult
 
-newtype MinimalImportsResult = MinimalImportsResult
-  {getMinimalImportsResult :: [(LImportDecl GhcRn, Maybe T.Text)]}
+data MinimalImportsResult = MinimalImportsResult
+  { forLens        :: [(Range, Int)]
+  , forCodeActions :: RM.RangeMap (Range, Int)
+  , forResolve     :: IM.IntMap (Range, T.Text) }
 
 instance Show MinimalImportsResult where show _ = "<minimalImportsResult>"
 
 instance NFData MinimalImportsResult where rnf = rwhnf
 
+data EIResolveData = ResolveOne
+                      { uri :: Uri
+                      , int :: Int }
+                    | ResolveAll
+                      { uri :: Uri }
+                    deriving (Generic, ToJSON, FromJSON)
+
 exportedModuleStrings :: ParsedModule -> [String]
 exportedModuleStrings ParsedModule{pm_parsed_source = L _ HsModule{..}}
   | Just export <- hsmodExports,
     exports <- unLoc export
     = map (T.unpack . printOutputable) exports
 exportedModuleStrings _ = []
 
-minimalImportsRule :: Recorder (WithPriority Log) -> Rules ()
-minimalImportsRule recorder = define (cmapWithPrio LogShake recorder) $ \MinimalImports nfp -> do
+minimalImportsRule :: Recorder (WithPriority Log) -> (ModuleName -> Bool) -> Rules ()
+minimalImportsRule recorder pred = define (cmapWithPrio LogShake recorder) $ \MinimalImports nfp -> do
   -- Get the typechecking artifacts from the module
-  tmr <- use TypeCheck nfp
+  Just tmr <- use TypeCheck nfp
   -- We also need a GHC session with all the dependencies
-  hsc <- use GhcSessionDeps nfp
+  Just hsc <- use GhcSessionDeps nfp
   -- Use the GHC api to extract the "minimal" imports
-  (imports, mbMinImports) <- liftIO $ extractMinimalImports hsc tmr
+  (imports, Just mbMinImports) <- liftIO $ extractMinimalImports (Just hsc) (Just tmr)
   let importsMap =
         Map.fromList
           [ (realSrcSpanStart l, printOutputable i)
-            | L (locA -> RealSrcSpan l _) i <- fromMaybe [] mbMinImports
-            , not (isImplicitPrelude i)
+            | L (locA -> RealSrcSpan l _) i <- mbMinImports
+            -- Don't we already take care of this with the predicate?
+            --, not (isImplicitPrelude i)
           ]
       res =
-        [ (i, Map.lookup (realSrcSpanStart l) importsMap)
-          | i <- imports
-          , RealSrcSpan l _ <- [getLoc i]
+        [ (realSrcSpanToRange l, minImport)
+          | i@(L (locA -> src) imp) <- imports
+          , not (isQualifiedImport imp)
+          , not (isExplicitImport imp)
+          , let L _ mn = ideclName imp
+          , pred mn
+          , let RealSrcSpan l _ = getLoc i
+          , Just minImport <- [Map.lookup (realSrcSpanStart l) importsMap]
         ]
-  return ([], MinimalImportsResult res <$ mbMinImports)
-  where
-    isImplicitPrelude :: (Outputable a) => a -> Bool
+  uniques <- liftIO $ replicateM (length res) (U.hashUnique <$> U.newUnique)
+  let uniqueAndRangeAndText = zip uniques res
```

Comment:
I sometimes write this sort of thing like this:
```
uniqueAndRangeAndText = for res $ \r ->
  u <- U.hashUnique <$> U.newUnique
  pure (r, u)
```
I just find it a bit easier often sine you have a variable for the thing and a natural place for a do block